---
title: "Project_Week4_StatisticsGenomicDataScience"
author: "Haris Spyridis"
date: "2024-04-17"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


#Question 1
Question 1

When performing gene set analysis it is critical to use the same annotation as was used in pre-processing steps. Read the paper behind the Bottomly data set on the ReCount database: http://www.ncbi.nlm.nih.gov/pubmed?term=21455293

Using the paper and the function: supportedGenomes()supportedGenomes() in the goseq package can you figure out which of the Mouse genome builds they aligned the reads to.
```{r}
library(goseq)

supportedGenomes()
```




#Question 2
Load the Bottomly data with the following code and perform a differential expression analysis using limmalimma with only the strain variable as an outcome. How many genes are differentially expressed at the 5% FDR level using Benjamini-Hochberg correction? What is the gene identifier of the first gene differentially expressed at this level (just in order, not the smallest FDR) ? (hint: the featureNamesfeatureNames function may be useful)


```{r}
#load the data
library(Biobase)
library(limma)
library(genefilter)
library(goseq)
library(org.Mm.eg.db)

con =url("http://bowtie-bio.sourceforge.net/recount/ExpressionSets/bottomly_eset.RData")
load(file=con)
close(con)
bot = bottomly.eset
pdata_bot=pData(bot)
fdata_bot = featureData(bot)
edata = exprs(bot)
fdata_bot = fdata_bot[rowMeans(edata) > 5]
edata = edata[rowMeans(edata) > 5, ]
edata = log2(edata+1)
head(edata)
```

```{r}
#fit a model using limma package
mod = model.matrix(~pdata_bot$strain)
lm = lmFit(edata, mod)
ebays = eBayes(lm)

#get the P values for limma package (it uses Bh by deafult)
limma_outp = topTable(ebays, 
                     number = dim(edata)[1],
                     sort= "none")

limma_outp[limma_outp$adj.P.Val<0.05,]
sum(limma_adj$adj.P.Val<0.05)
limma_outp[limma_outp$adj.P.Val<0.05,]

#check for any Na
any(is.na(limma_outp))

#binary vector of genes that have adju.p.val <0.05
genes_de = as.integer(limma_outp$adj.P.Val<0.05)
#give names to the vector in order to be used in the goseq function (needs vector 0/1 together with names of genes)
names(genes_de) = rownames(edata)


```



#Question 3:
Use the nullp and goseq functions in the goseqgoseq package to perform a gene ontology analysis. What is the top category that comes up as over represented? (hint: you will need to use the genome information on the genome from question 1 and the differential expression analysis from question 2.

```{r}
?nullp
#prob weight function 
prob_weight_func = nullp(genes_de, "mm9", "ensGene")

#retrive the annotations
GO.wall = goseq(prob_weight_func, "mm9", "ensGene")

head(GO.wall)
```

#Question 5:
Load the Bottomly data with the following code and perform a differential expression analysis using limma  and treating strain as the outcome but adjusting for lane as a factor. Then find genes significant at the 5% FDR rate using the Benjamini Hochberg correction and perform the gene set analysis with goseqgoseq following the protocol from the first 4 questions. How many of the top 10 overrepresented categories are the same for the adjusted and unadjusted analysis?


```{r}
mod5_lane = model.matrix(~pdata_bot$strain + as.factor(pdata_bot$lane.number))
lm_adj = lmFit(edata, mod5_lane)
ebaye5_adj = eBayes(lm_adj)

#get the adjusted P-values
limma_adj_outp = topTable(ebaye5_adj, 
                     number = dim(edata)[1],
                     sort = "none")
sum(limma_adj_outp$adj.P.Val<0.05)
genes_de_adj = as.integer(limma_adj_outp$adj.P.Val<0.05)

names(genes_de_adj) = rownames(edata)

pwf_5_adj = nullp(genes_de_adj, "mm9", "ensGene")
GO.wall5_adj = goseq(pwf_5_adj, "mm9", "ensGene")


length(intersect(GO.wall$category[1:10], GO.wall5_adj$category[1:10]))

```
